### PR TITLE
[SPARK-34265][WIP][PYTHON][SQL] Instrument Python UDF execution using SQL Metrics

### PR DESCRIPTION
**Pull request by LucaCanali**
### What changes were proposed in this pull request?
This proposes to add SQLMetrics instrumentation for Python UDF execution.  
The proposed metrics with a short description are:  

<code>time spent executing</code> time spent executing the Python UDF and sending the results back to the executors. It can contain a component related to the time spent waiting for data to process, when data sending is slower compared to execution time
<code>time spent sending data</code> time spent sending data to the Python workers, this is part of the Python UDF execution time
<code>time spent sending code</code> time spent sending Python code to the Python workers
<code>bytes of code sent</code> the number of bytes of serialized Python code sent the to the Python workers 
<code>bytes of data returned</code> the number of bytes of serialized data received back from the Python workers 
<code>bytes of data sent</code> the number of bytes of serialized data sent the to the Python workers
<code>number of batches returned</code> the number of data batches received back from the Python workers
<code>number of batches processed</code> the number of data batches sent to the Python workers
<code>number of rows returned</code> the number of rows returned by the Python workers
<code>number of rows processed</code> the number rows sent to the Python workers 

### Why are the changes needed?
This aims at improving monitoring and performance troubleshooting of Python code called by Spark, via UDF, Pandas UDF or with MapPartittions. 

### Does this PR introduce _any_ user-facing change?
The introduced SQL metrics are exposed to the end users via the WebUI interface, in the SQL tab, for execution steps related to Python UDF execution, namely BatchEvalPython, ArrowEvalPython, AggregateInPandas, FlaMapGroupsInPandas, FlatMapsCoGroupsInPandas, MapInPandas, WindowsInPandas.
See also the screenshot with the metrics introduced:

### How was this patch tested?
Manually tested + a python test has been added.

Example code used for testing:
```
from pyspark.sql.functions import col, pandas_udf
import time

@pandas_udf("long")
def test_pandas(col1):
  time.sleep(0.02)
  return col1 * col1

spark.udf.register("test_pandas", test_pandas)
spark.sql("select rand(42)*rand(51)*rand(12) col1 from range(10000000)").createOrReplaceTempView("t1")
spark.sql("select max(test_pandas(col1)) from t1").collect()
```

This is used to test with more data pushed to the Python workers
```
from pyspark.sql.functions import col, pandas_udf
import time

@pandas_udf("long")
def test_pandas(col1,col2,col3,col4,col5,col6,col7,col8,col9,col10,col11,col12,col13,col14,col15,col16,col17):
  time.sleep(0.02)
  return col1


spark.udf.register("test_pandas", test_pandas)
spark.sql("select rand(42)*rand(51)*rand(12) col1 from range(10000000)").createOrReplaceTempView("t1")
`spark.sql("select max(test_pandas(col1,col1+1,col1+2,col1+3,col1+4,col1+5,col1+6,col1+7,col1+8,col1+9,col1+10,col1+11,col1+12,col1+13,col1+14,col1+15,col1+16)) from t1").collect()
```

### CODE REVIEW

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
cc: @HyukjinKwon @ueshin

**Comment by HyukjinKwon** [comment]
cc @BryanCutler too

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowPythonRunner.scala [hunk 1]
```diff
@@ -72,6 +83,9 @@ class ArrowPythonRunner(
         }
 
         PythonUDFRunner.writeUDFs(dataOut, funcs, argOffsets)
+        val deltaTime = System.nanoTime()-startTime
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowPythonRunner.scala:86]: nit: style: need spaces before and after `-`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowPythonRunner.scala [hunk 2]
```diff
@@ -72,6 +83,9 @@ class ArrowPythonRunner(
...
+        pythonCodeSerializeTime += deltaTime
+        pythonCodeSent += dataOut.size()
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowPythonRunner.scala:88]: I guess we also need to use a delta size between before and after the code is sent?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowPythonRunner.scala [hunk 3]
```diff
@@ -88,14 +102,22 @@ class ArrowPythonRunner(
...
             writer.writeBatch()
+
+            val deltaTime = System.nanoTime() - startTime
+            pythonSerializeTime += deltaTime
+            val rowCount = root.getRowCount
+            pythonNumBatchesSent += 1
+            pythonNumRowsSent += rowCount
             arrowWriter.reset()
           }
+          pythonDataSent += dataOut.size()
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowPythonRunner.scala:120]: ditto.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonExec.scala [hunk 4]
```diff
@@ -83,12 +97,19 @@ case class BatchEvalPythonExec(udfs: Seq[PythonUDF], resultAttrs: Seq[Attribute]
...
         // fast path for single UDF
         mutableRow(0) = fromJava(result)
+        val deltaTime = System.nanoTime() - startTime
+        pythonExecTime += deltaTime
         mutableRow
       } else {
-        fromJava(result).asInstanceOf[InternalRow]
+        val res = fromJava(result).asInstanceOf[InternalRow]
+        val deltaTime = System.nanoTime() - startTime
+        pythonExecTime += deltaTime
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonExec.scala:111]: Is this really calculating the exec time? Seems like it's only calculating the time for `fromJava(result)`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala [hunk 5]
```diff
@@ -71,10 +77,19 @@ private[python] trait PythonArrowOutput { self: BasePythonRunner[_, ColumnarBatc
         }
         try {
           if (reader != null && batchLoaded) {
+            val bytesReadStart = reader.bytesRead()
+            val startTime = System.nanoTime()
             batchLoaded = reader.loadNextBatch()
+            val deltaTime = System.nanoTime() - startTime
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala:83]: Same here? Seems like it's just calculating the batch loading time?

File: docs/web-ui.md [hunk 6]
```diff
@@ -404,6 +404,16 @@ Here is the list of SQL metrics:
...
+<tr><td> <code>number of batches returned</code> </td><td> the number of data batches received back from the Python workers </td><td> BatchEvalPython, ArrowEvalPython, AggregateInPandas, FlaMapGroupsInPandas, FlatMapsCoGroupsInPandas, MapInPandas, WindowsInPandas </td></tr>
+<tr><td> <code>number of batches processed</code> </td><td> the number of data batches sent to the Python workers </td><td> BatchEvalPython, ArrowEvalPython, AggregateInPandas, FlaMapGroupsInPandas, FlatMapsCoGroupsInPandas, MapInPandas, WindowsInPandas </td></tr>
+<tr><td> <code>number of rows returned</code> </td><td> the number of rows returned by the Python workers </td><td> BatchEvalPython, ArrowEvalPython, AggregateInPandas, FlaMapGroupsInPandas, FlatMapsCoGroupsInPandas, MapInPandas, WindowsInPandas </td></tr>
+<tr><td> <code>number of rows processed</code> </td><td> the number rows sent to the Python workers </td><td> BatchEvalPython, ArrowEvalPython, AggregateInPandas, FlaMapGroupsInPandas, FlatMapsCoGroupsInPandas, MapInPandas, WindowsInPandas </td></tr>
```
Comment (maropu) [inline docs/web-ui.md:416]: For reviews, could you put the screenshot of the updated doc in the PR description?

File: python/pyspark/sql/tests/test_pandas_sqlmetrics.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            return col1 * col1
+
+        res = self.spark.range(10).select(test_pandas("id")).collect()
+
+        statusStore = self.spark._jsparkSession.sharedState().statusStore()
+        executionMetrics = statusStore.execution(0).get().metrics().mkString()
+
+        for metric in PythonSQLMetrics:
+            self.assertIn(metric, executionMetrics)
+
```
Comment (maropu) [inline python/pyspark/sql/tests/test_pandas_sqlmetrics.py:56]: nit: one more blank line.

File: python/pyspark/sql/tests/test_pandas_sqlmetrics.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+from pyspark.sql.functions import pandas_udf
+from pyspark.testing.sqlutils import ReusedSQLTestCase, have_pandas, have_pyarrow, \
+    pandas_requirement_message, pyarrow_requirement_message
+
+
+@unittest.skipIf(
+    not have_pandas or not have_pyarrow,
+    pandas_requirement_message or pyarrow_requirement_message)  # type: ignore[arg-type]
+class PandasSQLMetrics(ReusedSQLTestCase):
```
Comment (maropu) [inline python/pyspark/sql/tests/test_pandas_sqlmetrics.py:28]: Could you add tests in a Scala side (e.g., `SQLMetricsSuite`), too?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonSQLMetrics.scala [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.python
+
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.metric.SQLMetrics
+
+private[python] trait PythonSQLMetrics extends SparkPlan {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonSQLMetrics.scala:23]: These plans already extend `SparkPlan` (e.g., https://github.com/apache/spark/pull/31367/files#diff-ce3497483dcd52a600e1414d36f15e96f766ea94f3c50f075926716b7e42c724R57), so could you use self-type instead?

```
private[python] trait PythonSQLMetrics {
  self: SparkPlan
...
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonSQLMetrics.scala [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  override val metrics = Map(
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonSQLMetrics.scala:24]: If you want to implement many metrics, how about using `Map[String, SQLMetric]` instead? (see `ShuffledRowRDD`)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDFRunner.scala [hunk 11]
```diff
@@ -31,10 +32,17 @@ import org.apache.spark.sql.internal.SQLConf
...
+    pythonDataSerializeTime: SQLMetric,
+    pythonCodeSerializeTime: SQLMetric,
+    pythonCodeSent: SQLMetric,
+    pythonDataReceived: SQLMetric,
+    pythonDataSent: SQLMetric,
+    pythonNumBatchesReceived: SQLMetric,
+    pythonNumBatchesSent: SQLMetric)
   extends BasePythonRunner[Array[Byte], Array[Byte]](
     funcs, evalType, argOffsets) {
-
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDFRunner.scala:37]: nit: revert the unnecessary change

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala [hunk 12]
```diff
@@ -26,16 +26,22 @@ import org.apache.arrow.vector.VectorSchemaRoot
...
 import org.apache.spark.sql.util.ArrowUtils
 import org.apache.spark.sql.vectorized.{ArrowColumnVector, ColumnarBatch, ColumnVector}
 
-/**
- * A trait that can be mixed-in with [[BasePythonRunner]]. It implements the logic from
- * Python (Arrow) to JVM (ColumnarBatch).
- */
-private[python] trait PythonArrowOutput { self: BasePythonRunner[_, ColumnarBatch] =>
+// An abstract class implementing the logic from Python (Arrow) to JVM (ColumnarBatch).
+private[python] abstract class PythonArrowOutput[IN](
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala:36]: Could you avoid this change? How about writing it like this?
```
private[python] trait PythonArrowOutput { self: BasePythonRunner[_, ColumnarBatch] =>
  def pythonDataReceived: SQLMetric
  def pythonExecTime: SQLMetric
  def pythonNumRowsReceived: SQLMetric
  def pythonNumBatchesReceived: SQLMetric
  ...
```
Then, can you override them in the `ArrowPythonRunner` side?

**Review by maropu — COMMENTED** [review]
Adding new metrics for python-related plan nodes looks useful (I think we need to discuss which metrics to include and which metrics to not include though)

**Comment by maropu** [comment]
Could you list up the new metrics (their names and description) that this PR intends to add in the `What changes were proposed in this pull request?` section?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonExec.scala (hunk 4 above)
Comment (LucaCanali) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonExec.scala:111]: Correct, this is just a component, actually small and could be omitted. The most important part of the execution for BatchEvalPythonExec time is measured in PythonUDFRunner.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala (hunk 5 above)
Comment (LucaCanali) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala:83]: Good point. From what I can see by experimenting, is that this actually triggers execution, so the time measured here seems to be a relevant value for the execution time. Ideally we would like to measure on the Python side, although this seems good enough?

**Comment by LucaCanali** [comment]
I have updated PR with the proposed list of metrics. The list currently contains 10 metrics.

I can see the need to have just a few important metrics.
However, I can also see that the Python UDF implementation is complex with many moving parts, metrics can help with troubleshooting with corner cases too, in those circumstances more metrics mean more flexibility and more chances to find where the root cause of the problem is.


Another point for discussion is how accurate are the metrics in the current implementation? I have run a few tests to check that the values measured make sense and are in the ballpark of what expected. 
In particular, measuring execution time can be challenging at times, as with this we attept to do it from JVM. I have put in the metrics description some hints to the nuances I found when testing . Send and receive time back and forth between JVM and Python workers seem to overlap in some cases, due the use of queues. 

I think the "time spent sending data" can be useful when troubleshooting cases where the performance problem is with sending a lot of data to Python. Time spent executinf is probably the key metric to understand the overall performance. Number of rows returned will probably be another useful one, to understand the progress of the execution when monitoring the progress of an active query.

**Comment by maropu** [comment]
> I have updated PR with the proposed list of metrics. The list currently contains 10 metrics.

Could you use texts to list up the metrics  instead of the image? That's because the PR description will appear in the commit logs after this PR merged and the image data makes it hard to search commit logs, I think.

File: python/pyspark/sql/tests/test_pandas_sqlmetrics.py (hunk 8 above)
Comment (LucaCanali) [reply python/pyspark/sql/tests/test_pandas_sqlmetrics.py:28]: Good point. I am now sure on how to do that for Python UDF though.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/CoGroupedArrowPythonRunner.scala [hunk 13]
```diff
@@ -71,6 +83,9 @@ class CoGroupedArrowPythonRunner(
         }
 
         PythonUDFRunner.writeUDFs(dataOut, funcs, argOffsets)
+        val deltaTime = System.nanoTime()-startTime
+        pythonCodeSerializeTime += deltaTime
+        pythonCodeSent += dataOut.size()
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/CoGroupedArrowPythonRunner.scala:88]: ditto as https://github.com/apache/spark/pull/31367#discussion_r566439263

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/CoGroupedArrowPythonRunner.scala [hunk 14]
```diff
@@ -83,6 +98,7 @@ class CoGroupedArrowPythonRunner(
           writeGroup(nextRight, rightSchema, dataOut, "right")
         }
         dataOut.writeInt(0)
+        pythonDataSent += dataOut.size()
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/CoGroupedArrowPythonRunner.scala:101]: ditto.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDFRunner.scala [hunk 15]
```diff
@@ -46,12 +54,23 @@ class PythonUDFRunner(
     new WriterThread(env, worker, inputIterator, partitionIndex, context) {
 
       protected override def writeCommand(dataOut: DataOutputStream): Unit = {
+        val startTime = System.nanoTime()
         PythonUDFRunner.writeUDFs(dataOut, funcs, argOffsets)
+        val deltaTime = System.nanoTime()-startTime
+        pythonCodeSerializeTime += deltaTime
+        pythonCodeSent += dataOut.size()
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDFRunner.scala:61]: ditto.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonSQLMetrics.scala (hunk 10 above)
Comment (LucaCanali) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonSQLMetrics.scala:25]: My original thought here was to use a trait so that we can reuse the metrics definitaion with the many physical plans that need it (BatchEvalPythonExec, ArrowEcalPythonExec, AggregateInPandasExec, etc). Other implementations of this can work too. I see that `ShuffleRowRDD` uses a custom object `SQLShuffleMetricsReporter` to define and handle the metrics. Can you please elaborate a bit more on your proposal?

**Comment by LucaCanali** [comment]
This has been blocked for a while. I believe that from the discussion it has emerged:
- the proposed implementation in this PR possible has too many metrics, which could be confusing
- the timing metrics are hard to implement and to understand if the proposed implementation is correct/useful

I propose to take some of the ideas in the PR to a different and less complex PR: #33559
